RCL

Kurfursten opened the pull request:
Rapid Cable Layer (RCL) allows you to quickly smooth-wire en masse.

![cables](https://cloud.githubusercontent.com/assets/9782036/8096339/c613404a-0f9b-11e5-8381-ddf9190a6b6e.png)

Video (one day) here: http://a.uguu.se/hxljhg.webm

In addition to the above sprites, I have added for the icon_state 2 additional icon_states that show the cable coil 2/3 full and 1/3 full respectively. You can see the icon changing in the video.

Now fits on the CE belt!
- Needs Changelog
- Special thanks to anonymous spriter from thread

PJB3005 commented:
3,477  vgstation13.dme
1,739 additions, 1,738 deletions not shown

Sounds legit

otherwise :+1: though

Kurfursten commented:
@PJB3005 I had to include the dme in the commit because I added a new file (rcl.dm), the fact that it's going nuts with changes is troubling, though, good catch

Zuhayr commented:
Run the maptools in the tools/mapmerge folder.

Probe1 commented:
I'd say one in each electrical locker and add it to the hacked autolathe.  Possibly a crate orderable from cargo as well.

Kurfursten commented:
@Zuhayr It doesn't include any map changes, though (?)

Zuhayr commented:
Yeah I deleted my comment, I'm a moron who can't read.

PJB3005 commented:
EOLs popping their head up if I had to guess.

FudgePucker commented:
@Kurfursten 
I know I might be annoying and obnoxious with idea suggestions and irrelevant comments in your PR's,
but I just wanted to say I love you dude.
You are just super incredible.
Also can you give one of these to cyborgs and MoMMIs? Or at least an upgrade?

Kurfursten commented:
Thanks. Might not be a bad idea for an engineering cyborg upgrade.  I'm hesitant to give more toys to MoMMIs because they're so great at engineering already.

clusterfack commented on code/game/objects/items/weapons/RCL.dm:
```diff
@@ -0,0 +1,205 @@
+
+/obj/item/weapon/rcl/update_icon()
+	switch(cables)
+		if(21 to INFINITY)
```
Comic is going to tell you these are magic numbers, probably

Kurfursten replied:
I think Comic is the one who told me to do it this way originally when I did a similar spriting thing for soldering iron

clusterfack commented on code/game/objects/items/weapons/RCL.dm:
```diff
@@ -0,0 +1,205 @@
+
+/obj/item/weapon/rcl/examine(mob/user)
+	..()
+	user << "<span class='info'>It contains [cables]/30 cables.</span>"
```
You should add a max cables variable for adminbus, instead of hardcoding it at 30

AutismSpider commented:
Add it mommis and Engi borgs

Mommi RCL is 30/40 cables Engi borg/human RCL is 60

IratePirate commented:
>add it mommis

I'd avoid giving it to MoMMIs at all, it's not like they need yet more "buffs".

PJB3005 commented:
Eh MoMMIs can steal their own.

ComicIronic commented:
Make it twohanded to stop MoMMis using it.

And don't copypaste code.

RemieRichards commented:
Just store an internal cable coil object and call the procs from it instead of copypasting like a madman on steroids.

ComicIronic commented:
Exactly that.

FudgePucker commented:
>Make it twohanded to stop MoMMIs using it.

Excuse me if I am wrong but this seems like the most retarded solution to the problem.
Even making it so that MoMMIs can't use it like with guns would be a better solution.

How would you two hand it?
Clicking on the device might make it so that cable coils fall out of it.
That would be a lot better feature than a MoMMI nerf.
>Oh man I wish I could get some cable coils out of my RCL to finish up this airlock/APC/air alarm, but at least MoMMIs can't use it.

Kurfursten commented:
@RemieRichards Bleh, I'm going to have to rewrite cable coil in that case. I made a new proc because cable coil itself copy/pastes its own code

RemieRichards commented:
@Kurfursten Unless you've made some rather... *odd* decisions, I highly doubt you'd need to rewrite cable coils, meaning it'd be your own stuff you should refactor.

Kurfursten commented:
@FudgePucker attack self is used as the placement proc, ejecting wires would have to be a verb

Kurfursten commented:
@RemieRichards You're right, I don't NEED to rewrite it, but only in the sense that I don't NEED to rewrite my own code to remove copy/pasting. I'm just bellyaching, it'll get done.

PJB3005 commented:
@Kurfursten OH GOD NOT VERBS

Iamgoofball commented:
needs a rapid rapid dispenser to be added

The RRD

Kurfursten commented:
@Iamgoofball Major missed opportunity by not calling the tile painter the "rapid tile painter"

Kurfursten commented:
**SUMMARY OF CHANGES**

* Now we internally store a cable coil
* Rewrote cable coil partially to use the turf_place proc more effectively
* Now has an eject cables verb
* Fixed a minor bug that none of you knew about

ComicIronic commented:
VERBS?

Why not just snip it out with wirecutters?

Probe1 commented:
w-why not just make it a container

PJB3005 commented:
OR use contextual clicking thingy comic made?

Kurfursten commented:
@Probe1 attack self is already being used, in response to both being a container and wielded

@ComicIronic could you elaborate on why it's better to restrict wire removal behind a tool wall?

ComicIronic commented:
It's hardly a tool wall, since anyone laying cable will have wirecutters to fix mistakes and will be an engineer with a pair anyways.

And verbs are awful, that's why. They clunkify the interaction, lag up the game, and are generally used as a poor man's substitute for better options in terms of using items or machines.

DrCelt commented:
The RCL should not be prohibited from MoMMIs as @ComicIronic has suggested. Along with a number of practical reasons why that'd be silly, the overwhelming majority of people polled one round were against prohibition. !(http://i.imgur.com/VbI74Is.png)

Kurfursten commented:
**SUMMARY OF CHANGES**
* You can now release cables from the RCL with the screwdriver (and cartridges from the RCD with the same).
* No more eject verb
* MoMMIs can now pick up the RCL
* Icon conflict resolved
* Can now be placed in the CE belt

It's ready for NFD

ComicIronic commented:
1800 deletions

ComicIronic commented on code/modules/power/cable.dm:
```diff
@@ -182,6 +182,10 @@ By design, d1 is the smallest direction and d2 is the highest
 		coil.cable_join(src, user)
+	else if(istype(W, /obj/item/weapon/rcl))
+		var/obj/item/weapon/rcl/R = W
+		R.loaded.cable_join(src, user)
```
if R.loaded

You're going to runtime here.

ComicIronic commented on code/game/objects/items/weapons/RCL.dm:
```diff
@@ -0,0 +1,97 @@
+/obj/item/weapon/rcl/attackby(obj/item/weapon/W, mob/user)
+	if(isscrewdriver(W))
+		user << "<span class='notice'>You loosen the securing screws on the side, allowing you to lower it and retrieve the wires.</span>"
+		loaded.loc = usr.loc
```
usr

PJB3005 replied:
`forceMove()` too!

ComicIronic replied:
Well, it doesn't have to be forceMove unless it's a mob or it has luminosity.

PJB3005 replied:
Assuming something doesn't have luminosity is shit code, you shouldn't ever assume that, even if it's just for the adminbus

ComicIronic commented on code/game/objects/items/weapons/RCL.dm:
```diff
@@ -0,0 +1,97 @@
+	update_icon()
+	if(!loaded.amount)
+		user << "<span class='notice'>The last of the cables unreel from the [src].</span>"
+		loaded = null
```
You should return the cable coil to the object pool.

ComicIronic commented on code/game/objects/items/weapons/RCL.dm:
```diff
@@ -0,0 +1,97 @@
+		icon_state = "rcl-0"
+		item_state = "rcl-0"
+		return
+	switch(loaded.amount)
```
A neat trick you might prefer is to instead do
icon_state = "rcl-[loaded.amount - (loaded.amount % 10)]"
Which will floor it to the nearest multiple of 10 for you.

Kurfursten replied:
This doesn't work in this base because I actually want to ceiling and if I use the opposite (10- loaded.amount mod 10) it would round 20 to 30 and 0 to 10

ComicIronic replied:
Then you can do loaded.amount + [10 - (loaded.amount % 10)]

Kurfursten replied:
That doesn't really work perfectly for the reasons I said above. Take 0 case for example.

0 + (10 - 0 % 10) = 10, therefore it would never show as empty

ComicIronic commented on code/game/objects/items/weapons/RCL.dm:
```diff
@@ -0,0 +1,97 @@
+
+/obj/item/weapon/rcl/examine(mob/user)
+	..()
+	user << "<span class='info'>It contains [loaded.amount]/30 cables.</span>"
```
Runtime here is loaded is null.

ririchiyo commented:
generally you just want to manually add your file to the .DME than allowing dream maker to ever modify your files.

SkowronX commented:
Do not merge. A bunch of the icons are somehow missing from the inhand dmis.

Kurfursten commented:
@SkowronX Could you be more specific?

PJB3005 commented:
Is the file read only, if so, you're opening a file that's got > 512 icons and DM automatically fucks up the file if you were to save it.

Kurfursten commented:
Not much to be done about it now, officially on vacation, will pick it up in a week if it's not working
